open-company/open-company-web
# REFACTOR: Move the activities related actions to the proper namespace

Step 1 of the refactor https://trello.com/c/nOF8CduC

Remove all the confused actions from the shared namespace and move them to the dedicated namespace in form of action creators.
The following actions have been moved:

- `:entry`
- `:entry-clear-local-cache`
- `:entry-save`
- `:entry-save/finish`
- `:entry-save/failed`
- `:entry-publish`
- `:entry-publish-with-board/finish`
- `:entry-publish/finish`
- `:entry-publish/failed`
- `:activity-delete`
- `:activity-delete/finish`
- `:activity-share-show`
- `:activity-share-hide`
- `:activity-share-reset`
- `:activity-share`
- `:activity-share/finish`
- `:secure-activity-get`
- `:activity-get/finish`
Also moved:
- `:all-posts-get`
- `:all-posts-get/finish`
- `:all-posts-more`
- `:all-posts-more/finish`

Delete:
- `:activity-move`
Also deleted:
-  `:all-posts-calendar`
-  `:calendar-get`
- ` :calendar-get/finish`

### Tests

Setup your environment: use an org with at least 2 sections and 2 users (both admins is ok). For the purpose of these tests we will call org the org, S1 and S2 the sections and U1 and U2 the users.

#### Testing `:entry`.
Retrieve a post from the server, it's used in 2 places, when a comment is added from another session (a ws interaction message is received), when a reaction is added from another session (a ws interaction message is received), when the user reacts picking a reaction from EmojiMart (it refreshes the complete entry to retrieve the new links)

Test 1:
- open S1 with U1 and U2
- U1 add a comment
- check if U2 refreshes the post

Test 2:
- open S1 with U1 and U2
- U1 add a reaction via the picker
- check if U1 refreshes the post
- check if U2 refreshes the post

#### Testing `:entry-clear-local-cache`.
Clean the local indexedDB from the edited post data. Used in 2 places: when creating a new post (or editing a draft) and when editing a published post.

Setup:
- open a session with U1 on B1
- go to Dev Tools -> Application -> indexedDB -> localforage -> keyvaluepairs
- click the 🚫icon saying `Clear Object Store`

Test 1:
- U1 compose a new post
- add some editing
- wait 5 seconds
- go to Dev Tools -> Application -> indexedDB -> localforage -> keyvaluepairs
- click the refresh icon (besides the 🚫)
- [x] do you see one record? Good
- U1 dismiss editing
- refresh the indexedDB again
- [x] do you see 0 records? Good

Test 2:
- repeat Test 1 but this time editing a draft

Test 3:
- repeat Test 1 but this time editing a post

#### Testing `:entry-save`.
Creates or saves a draft. Also this consequently tests `:entry-save/finish` and `:entry-save/failed`.

Test 1:
- U1 click Compose
- add headline and body
- click Save to draft
- [x] are you redirected to the drafts board? Good
- [x] do you see the newly added draft with all the data? Good
- click edit on the draft
- change something
- click Save to draft again
- [x] are you redirected to the drafts board? Good
- [x] do you see the latest edits? Good

#### Testing `:entry-publish`.
Publish a new post. Consequently tests also `:entry-publish-with-board/finish`, `:entry-publish/finish` and `:entry-publish/failed`.

Test 1:
- U1 clicks Compose
- add headline and body
- click Post
- [x] can you see the newly added post? Good

Test 2:
- U1 clicks Compose
- add headline and body
- clicks on the section
- add a new section
- [x] are you redirected to the newly created section? Good
- [x] do you see the newly added post? Good

#### Testing `:activity-delete`.
Deletes a post. Consequently tests also `:activity-delete/finish`.

Test 1:
- U1 goes to AP stream
- clicks on the ellipses menu of a post
- then clicks Delete
- clicks yes on the prompt
- [x] do you see the post disappearing? Good

Test 2:
- repeat Test 1 but from grid view

Test 3:
- repeat Test 1 but from fullscreen post

#### Testing `:activity-share`.
Also `:activity-share/finish`, `:activity-share-show`, `:activity-share-hide` and `:activity-share-reset`.

The first 2 `:activity-share` and `:activity-share/finish` are used to trigger the server activity share and to  handle the response.
The remaining 3 are used for the UI: show and hide the UI and reset the UI data once a share took place.

Test 1:
- U1 go to stream view AP
- clicks on ellipses menu
- clicks on Share
- [x] do you see the correct post title at the top? Good
- clciks on URL
- copy the URL and save it somewhere
- [x] dismiss the view, did it work? Good
- scrolls down to another post
- clicks on the ellipsis menu
- clicks on Share
- [x] do you see the new post title at the top? Good
- clicks on URL
- copy the new URL
- [x] is it different then the previous one? Good
- scrolls some more
- clicks on the ellipses menu of another post
- clicks share
- [x] the title is ok? Good
- clicks on email
- compile the form and clicks Share
- [x] do you see the request to share and its data? Good

#### Testing `:secure-activity-get`.
Used to retrieve the data of a post when shared via secure URL.

Test 1:
- U2 open up one of the previous share URL
- [x] are you redirected to the fullscreen post and the url switched to our post URL? Good

Test 2:
- open an incognito browser session
- open the dev tools
- select the Network tab
- paste the share URL of the previous test in the browser address field
- [x] do you see the secure post request? Good
- [x] could you load the secure post? Good

#### Testing `:all-posts-get`.
Also `:all-posts-get/finish`. It retrieves the AP data from API, it's used by the main org-dashboard component to retrieve the AP data when it mounts.

When moving this action i found an problem of circular dependency touching the comment and reaction stores so i moved the defmethod of the websocket receiving messages inside the actions from the ws-interaction-client namespace.
The following tests 2, 3 and 4 include also this.

Test 1:
- U1 open All Posts of the org
- [x] do you see the data? Good

Test 2:
- U1 open AP in stream view
- U2 open AP in stream view
- U1 comment on the first post
- [x] U2 sees the comment? Good

Test 3:
- U1 open S1
- U2 open AP and scroll to the first post from S1
- U1 react to the first post
- [x] U2 sees the reaction? Good

Test 4:
- U1 remove the previous reaction and add another
- [x] U2 sees the reaction? Good

#### Testing `:all-posts-more`.
Also `:all-posts-more/finish`. It retrieves more data for AP. Used in AP on grid and stream view.

NB: Make sure that org has enough post to at least 3 pages of All Posts

Test 1:
- U1 open AP
- switch to stream view if not already
- scrolls down quickly to the bottom of the page
- [x] do you see the message at the bottom saying something like _Retrieving more data_? Good
- [x] more data appear? Good
- scrolls down more this time slowly
- [x] do you see the scroll bar changing size and moving when more data is loaded (even if you didn't reach the bottom)? Good

Test 2:
- U1 opens AP
- switch to grid view if not already
- scroll quickly to the bottom
- [x] do you see more data appear (this time no message)? Good
- now scrolls slowly down
- [x] do you see the scroll bar changing size and moving when more data is loaded (even if you didn't reach the bottom)? Good
